rclpy/rclpy/node.py

```diff
@@ -325,24 +330,56 @@ def declare_parameters(
         :raises: ParameterAlreadyDeclaredException if the parameter had already been declared.
         :raises: InvalidParameterException if the parameter name is invalid.
         :raises: InvalidParameterValueException if the registered callback rejects any parameter.
+        :raises: TypeError if any tuple in :param:`parameters` does not match the annotated type.
         """
         parameter_list = []
         descriptor_list = []
-        for parameter_tuple in parameters:
-            name = parameter_tuple[0]
-            assert isinstance(name, str)
-            # Get value from initial parameters, of from tuple if it doesn't exist.
-            if name in self._initial_parameters:
-                value = self._initial_parameters[name].get_parameter_value()
-            elif parameter_tuple[1] is None:
-                value = ParameterValue()
-            else:
-                value = parameter_tuple[1]
-            assert isinstance(value, ParameterValue)
-            descriptor = parameter_tuple[2]
-            if descriptor is None:
-                descriptor = ParameterDescriptor()
-            assert isinstance(descriptor, ParameterDescriptor)
+        for index, parameter_tuple in enumerate(parameters):
+            if len(parameter_tuple) < 1 or len(parameter_tuple) > 3:
+                raise TypeError(
+                    'Invalid parameter tuple length at index {index} in parameters list: '
+                    '{parameter_tuple}; expecting length between 1 and 3'.format_map(locals())
+                )
+
+            value = ParameterValue()
+            descriptor = ParameterDescriptor()
+
+            # Get the values from the tuple, checking its types.
+            # Use defaults if the tuple doesn't contain value and / or descriptor.
+            try:
+                name = parameter_tuple[0]
+                assert \
+                    isinstance(name, str), \
+                    (
+                        'First element {name} at index {index} in parameters list '
+                        'is not a str.'.format_map(locals())
+                    )
+
+                # Get value from initial parameters, of from tuple if it doesn't exist.
+                if name in self._initial_parameters:
+                    value = self._initial_parameters[name].get_parameter_value()
+                else:
+                    value = parameter_tuple[1]
+                assert \
+                    isinstance(value, ParameterValue), \
```

Comment:
Does this work if I do `('foo', 42)`? Or do you always have to do `('foo', ParameterValue(42))`?

Reply:
`('foo', 42)` won't work; `ParameterValue` is actually the [message](https://github.com/ros2/rcl_interfaces/blob/master/rcl_interfaces/msg/ParameterValue.msg) defined in `rcl_interfaces`, so you need to specify the type and set the value where it corresponds (see this [test](https://github.com/ros2/rclpy/pull/345/files#diff-960e40f792ccd92281ee56dfb65726bcR671) as an example).

Reply:
That's not very convenient, we have a separate class for that in C++ (separate from the message type) that can be conversion constructed from common parameter value types.

I guess that doesn't have to be improved in this pr, but it makes this method pretty ugly to use.

Reply:
That's indeed a reason to use `Parameter` directly for `declare_parameter` and `declare_parameters`, but we go back to the problem with the namespace in `declare_parameters` that we found out in the previous PR: https://github.com/ros2/rclpy/pull/325#discussion_r280233234.

I agree that it would be better to have a version that accepts a `Parameter` and a `ParameterDescriptor`, but I don't see a way of achieving it other than creating a copy of the parameter with the given namespace inside the method. Let's discuss this in the meeting, shall we?

Reply:
As discussed in the meeting, we shall move forward with this PR and introduce a way of converting basically any value to a `ParameterValue` message so that these methods are more convenient to use. One possibility is adding a new class (analogous to `Parameter`) that does the job.